added method add_area in class MusicBrainzClient (editing.py)

Prathamesh-Ghatole opened the pull request:
Added a new method "add_area" in the MusicBrainzClient class.

Prathamesh-Ghatole committed:
Updated .gitignore to ignore __pycache__/ .vscode/ env/ and venv/ virtual environment folders in project directory

Prathamesh-Ghatole committed:
added method add_area in class MusicBrainzClient (editing.py)

reosarevok commented:
This seems like a good start - it does the basic thing it needs to. Can you link to some of the test areas added?

You will want to expand this further (in further commits would probably be the clearest) to allow submitting ISO codes (rarely needed for cities but they might be for subdivisions), and to allow submitting URLs (since you will always want to at least include the relationship to Wikidata and Geonames).

Prathamesh-Ghatole commented:
Thanks for the review!
Here are a few links to new areas added using the bot:

https://test.musicbrainz.org/area/1add1711-78fb-4f22-b59d-a06dfae0d6e4
https://test.musicbrainz.org/area/43baecd4-e279-41a4-be2e-85004f67d1d3
https://test.musicbrainz.org/area/e70a5bea-86bb-4765-97b2-5ceec9887d06

Also, should I add the other parameters (ISO Codes, URLs, etc.) in this PR itself, or in a new PR?

reosarevok commented:
I'd do it in the same PR but on different commits :)

Prathamesh-Ghatole commented:
Noted 📝

Prathamesh-Ghatole commented:
FINALLY updated the function with 3 commits to first of all update the function to use the browser API (Requests) instead of the forms API to tackle Mechanize's lack of javascript support.

The next 2 commits add the functionality to include ISO 3166 codes and external links.

Test areas (respective to each commit):
- https://test.musicbrainz.org/area/29da46f6-bf12-40db-aa97-d5f49f8c23d5
- https://test.musicbrainz.org/area/1d7c60ea-7ed1-42ac-ae7d-730988ea970e
- https://test.musicbrainz.org/area/aa89d971-90b6-4fb5-a8f0-cd16378cc5f4 (Links to pineapple page on wikidata for testing)

reosarevok commented on editing.py:
```diff
@@ -187,6 +187,8 @@ def add_area(self, area, edit_note, auto=False):
             "edit-area.iso_3166_1.0": area["ISO_3166_1"],
             "edit-area.iso_3166_2.0": area["ISO_3166_2"],
             "edit-area.iso_3166_3.0": area["ISO_3166_3"],
+            "edit-area.url.0.text": area["ext_links"],
```
You're going to need to add two external links anyway, so it would make sense for this to support any number of links, probably passed as an array of objects? (or well, a list of dictionaries, I guess :) ).

I guess the same is true of ISO codes technically, but you're unlikely to need more than one in practice, so that one is less important.

Prathamesh-Ghatole replied:
Sounds reasonable :)

Will add it ASAP in a couple of days

Prathamesh-Ghatole commented:
Updated with the ability to pass multiple ISO codes and external links as a list :)

e.g.: https://test.musicbrainz.org/area/cac137dc-385f-4d93-9f70-43364483a2e4

mwiencek commented on editing.py:
```diff
@@ -175,6 +175,53 @@ def add_artist(self, artist, edit_note, auto=False):
+            "edit-area.iso_3166_1.0": area["ISO_3166_1"],
+            "edit-area.iso_3166_2.0": area["ISO_3166_2"],
+            "edit-area.iso_3166_3.0": area["ISO_3166_3"],
+            "edit-area.url.0.text": area["ext_links"],
```
Some naming suggestions:
 * keep property names lowercase, so iso not ISO
 * avoid abbreviations, so external_links not ext_links

Prathamesh-Ghatole replied:
Updated✅

mwiencek commented on editing.py:
```diff
@@ -175,6 +175,53 @@ def add_artist(self, artist, edit_note, auto=False):
+            ):
+                raise Exception(
+                    "Length of the list for url and link type id is not equal"
+                )
```
I'd suggest making `ext_links` a list of dicts, with `text` and `link_type_id` properties. Then remove `ext_link_type_id`. No need to check if they are the same length then.

Prathamesh-Ghatole replied:
Updated✅

mwiencek commented on editing.py:
```diff
@@ -175,6 +175,53 @@ def add_artist(self, artist, edit_note, auto=False):
+
+        for k, v in temp_dict.items():
+            for i, val in enumerate(v):
+                data_final[k.replace("0", str(i))] = val
```
`temp_dict` and `data_final` seem unnecessary, or at least kinda confusing -- can't we just remove the bogus '0' properties from `data_template`, and then loop over `area['ext_links']` to add the correct properties to `data_template` in the first place?

You can have a helper function that works for both ext_links and iso codes (dicts and strings). It will likely be useful for other entity types in the future.

Prathamesh-Ghatole replied:
Updated✅

Prathamesh-Ghatole commented:
@mwiencek Noted! Should be resolved soon

Prathamesh-Ghatole commented:
Refactored the method as per @mwiencek 's suggestions

mwiencek commented on editing.py:
```diff
@@ -175,6 +175,45 @@ def add_artist(self, artist, edit_note, auto=False):
+            "edit-area.comment": area["disambiguation"],
+            "edit-area.type_id": area["type"],
+            "edit-area.edit_note": edit_note,
+            "edit-area.iso_3166_1.0": area["iso_3166_1"],
+            "edit-area.iso_3166_2.0": area["iso_3166_2"],
+            "edit-area.iso_3166_3.0": area["iso_3166_3"],
+            "edit-area.url.0.text": area["external_links"]["text"],
+            "edit-area.url.0.link_type_id": area["external_links"]["type"],
```
This looks better, but the placeholder `.0.` properties still feel unnecessary. You have to loop over `template.items()` twice (once to remove `None` properties, and another to clobber the "0" properties).

Why not something like
```python
expand_field('iso_3166_1', 'iso_3166_1')
expand_field('external_links', 'url')
```

etc. and just check for `None` plus add the necessary properties in one go? With that in mind, `external_links` should be a list of dicts; right now it looks like you have a dict of lists?

mwiencek commented on editing.py:
```diff
@@ -175,6 +175,45 @@ def add_artist(self, artist, edit_note, auto=False):
+        }
+
+        def create_payload(template):
+            # remove null values from template
+            template = {k: v for k, v in template.items() if v is not None}
```
I'd rather we throw an error if certain basic required fields like `name` are `None`. (You'll have to check which other fields are required.)

Prathamesh-Ghatole commented:
Updated with refactored add_area() function
Also added exceptions if required areas are not passed to the function

e.g.: https://test.musicbrainz.org/area/971886c3-48d8-4d00-a1c8-442c6d48834a

reosarevok commented on editing.py:
```diff
@@ -181,36 +181,40 @@ def add_area(self, area, edit_note, auto=False):
 
         data_template = {
             "edit-area.name": area["name"],
+            # Why is disambiguation named "comment" on MB?
```
That used to be the old name of the field :) it's a "disambiguation comment" but I guess it was renamed to disambiguation because people were using it to write random comments? Dunno, it's an old thing :) But renaming the stuff in the database schema is not trivial and we haven't had a big reason to.

Also, just ask in the PR with a comment next time, rather than adding a comment to the code that now needs removing :D

Prathamesh-Ghatole replied:
I see :D
Nice to see such little nuggets coming up like fossils in archaeology lol

mwiencek replied:
But is there a reason someone using the bot API has to use "disambiguation" as the key?

I'm asking because if the keys in `area` matched the field names 1:1, we wouldn't need `data_template` at all. We could just do `create_payload('edit-area', area, required_fields)`. (And they mostly already do except for comment and type_id?)

Prathamesh-Ghatole replied:
> But is there a reason someone using the bot API has to use "disambiguation" as the key?
> 
> I'm asking because if the keys in `area` matched the field names 1:1, we wouldn't need `data_template` at all. We could just do `create_payload('edit-area', area, required_fields)`. (And they mostly already do except for comment and type_id?)

Resolved :D

reosarevok commented on editing.py:
```diff
@@ -181,36 +181,40 @@ def add_area(self, area, edit_note, auto=False):
-                if isinstance(v, list):
-                    for i, val in enumerate(v):
-                        payload[k.replace("0", str(i))] = str(val)
+            for k1, v1 in template.items():
```
Can we use more explicit names for the variables? I understand that it's supposed to be a generic key and value variable, is that really the best we can do though?

reosarevok commented on editing.py:
```diff
@@ -175,6 +175,58 @@ def add_artist(self, artist, edit_note, auto=False):
+            "edit-area.url": area["url"],
+        }
+
+        def create_payload(template):
```
Should this be defined outside of area? It might be useful elsewhere later on too and it's written in a generic way anyway. I'd also make it so that `required_fields` is a passed parameter which depends on the entity type (so, define it in `add_area`, and then pass it to `create_payload`).

Prathamesh-Ghatole replied:
Yes, definitely! I think the create payload function could be a nice addition to the overall arsenal of generic functions in this bot. But where to add the function, and how generic to make it is the question that I'd love to have a discussion on.

reosarevok replied:
Assuming it works there, I'd probably put it on top outside of `MusicBrainzClient` like the existing `album_to_form`.

Prathamesh-Ghatole replied:
Noted :)

Prathamesh-Ghatole commented:
Updated with the following features:
- Better-named variables with explanatory comments.
- Generalized create_payload() function to be a generic function for the MusicBrainz bot
- Added better docstrings and examples!

test area: https://test.musicbrainz.org/area/19b712e7-b900-4380-875b-1ca69170dc7d

reosarevok commented on editing.py:
```diff
@@ -217,9 +254,37 @@ def add_artist(self, artist, edit_note, auto=False):
+        Args:
+            area (dict): a dictionary containing the area's data (format below)
+            edit_note (str): edit note
+            auto (bool, optional): Unused, kept to maintain format. Defaults to False.
```
auto marks whether it should be entered as a votable edit or as an auto-edit. I probably would expect these to be auto-edits, but we'd need to decide.

Prathamesh-Ghatole replied:
Thanks, I'll update it in the PR :)

Prathamesh-Ghatole commented:
Refactored add_area() to remove the redundant data_template dictionary and updated create_payload() to include a 'prefix' parameter as per @mwiencek 's suggestion

test area: https://test.musicbrainz.org/area/96be00f5-3c05-4a94-92a9-a5df1e90499e

Prathamesh-Ghatole committed:
added env/ and venv/ to .gitignore

Prathamesh-Ghatole commented:
Made a tiny change in the .gitignore file to ignore env/ and venv/ folders as well. Rebased it to the start of the PR.

Everything else should be the same as my last comment on this PR

Prathamesh-Ghatole commented:
Squashed commits as requested by @mwiencek
